### ISSUE TITLE
Some `alpaka::math::MathStdLib` functions trigger `[-Werror,-Wdouble-promotion]` when not called with `double`

### ISSUE THREAD

**Issue by jkelling** [alpaka-group/alpaka#1414]
Using `alpaka::math::MathStdLib` functions `sin` and `cos` with an aruments of a different precision than `float` (e.g. `float`), and probably other triggers the following clang (Main/14) warning:
```
~alpaka/include/alpaka/math/sin/Traits.hpp:37:32:
error: implicit conversion increases floating-point precision: 'const float' to 'double' [-Werror,-Wdouble-promotion]
                    return sin(arg);
                           ~~~ ^~~
```

This is because, e.g., `sin` is not overloaded for different precisions. For `float` `sinf` should be used. (`sinl` for `long double` although this may not be supported by alpaka anyway.)

**Comment by psychocoderHPC** [comment]
Which backend is producing the error?

**Comment by jkelling** [comment]
Currently none that is upstream.

The CI for #1308 showed float-to-doubke promotion warnings, which wanted to addressed them, by using the requested type consistently. This gets blocked by `alpaka::math` functions prompting float arguments to a double return value.

**Comment by jkelling** [comment]
One could also say: "all backends implementing the math concept using `alpaka::math::MathStdLib`", though. This warning would be triggered in user code when the user wants to use float, and it would point to an actual problem.

**Comment by jkelling** [comment]
This problem is not limited to clang. With [gcc a different warning is raised](https://github.com/alpaka-group/alpaka/pull/1308/checks?check_run_id=3777533235):
```
/home/runner/work/alpaka/alpaka/include/alpaka/rand/RandAlpaka.hpp:122:76: error: implicit conversion from 'float' to 'double' to match other operand of binary expression [-Werror=double-promotion]
  122 |                             const T mag = sigma * sqrt(static_cast<T>(-2.) * log(u1));
      |                                                        ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
/home/runner/work/alpaka/alpaka/include/alpaka/rand/RandAlpaka.hpp:122:49: error: implicit conversion from 'float' to 'double' to match other operand of binary expression [-Werror=double-promotion]
  122 |                             const T mag = sigma * sqrt(static_cast<T>(-2.) * log(u1));
      |                                           ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/runner/work/alpaka/alpaka/include/alpaka/rand/RandAlpaka.hpp:122:49: error: conversion from 'double' to 'float' may change value [-Werror=float-conversion]
```
Note: this compuation should be perfromed in `T = float`.

**Comment by jkelling** [comment]
In both cases one could remove the warnings by `static_cast`ing, but this way one would silently perform a potentially more expensive calculation in `double` precision counter to the user's intent.

**Comment by jkelling** [comment]
This turn out to not be the underlying issue of what occurred in #1308: The implementations subject to this issue should not have been called in the first place.

This issue remains though, albeit with nothing currently triggering it.

@sbastrakov @psychocoderHPC should we add specializations for the fallback trait implementations for `float` and `long double` just in case this gets used somewhere in the future? Or should these fallback be removed? Are they intentionally used by any backend?

**Comment by bernhardmgruber** [comment]
I don't know if it helps, but the compiler warning is reasonable. `sqrt` is a single function accepting a `double`, so of course a `float` is promoted. You might want to use `std::sqrt`, which is an overloaded function with overloads for all floating point types.

**Comment by jkelling** [comment]
The issue is, that this code is part of [the trait definitions in `alpaka::math`](https://github.com/alpaka-group/alpaka/blob/develop/include/alpaka/math/sqrt/Traits.hpp#L35-L37), hence my asking if this should be fixed.

**Comment by bernhardmgruber** [comment]
I found it myself now. I thought you where talking about the implementations `XXXStdLib` implementations, to which my comment before applied.
You seem to be referring to the ADL fallback for types other than the ones handled by the backends. So for example if you implement your own complex number:
```c++
MyComplex<float> c;
alpaka::math::sqrt(acc, c);
```
You can use the alpaka functions and they will try to find an implementation of `sqrt` in the namespace of `MyComplex` via ADL. If you end up in this fallback, then you seem to not implement the `ConceptMathXXX`s.

**Comment by jkelling** [comment]
Indeed. The thing, that this fallback exists, but causes these up-conversion warnings because of the way it is currently implemented. I think, the fallbacks should either not exist, making your code fail to compile if `ConceptMathXXX`s are missing, or should keep take care of the float type.

**Comment by sbastrakov** [comment]
I vaguely remember encountering a similar thing for my forever-going complex number implementation. I think the `sqrt` piece linked by @jkelling above is generally incorrect and just happened to work before. It should be instead:
```cpp
using std::sqrt;
return sqrt(arg);
```
Will check in more detail, and how it actually comes to play with my complex PR, by Monday.

**Comment by jkelling** [comment]
The implementation there is not incorrect for `T = double`, it is just that the ADL always takes the C version of `sqrt`, i.e. `double sqrt(double)` in the global scope, leading to the cast.

I think your suggestion would defeat the purpose of this fallback, because then the fallback would just be `MathStdLib`. -- I think the purpose of this fallback may be to use compiler build-ins on some backends. Does someone know why they exist?

**Comment by jkelling** [comment]
> You seem to be referring to the ADL fallback for types other than the ones handled by the backends.

If you have your own types, outside of alpaka, why would you want to use `alpaka::math::` on them? Would you not have your own `sqrt` in your namespace and call that via `sqrt(varOfMyType)` (or by qualified name)?

**Comment by bernhardmgruber** [comment]
> Does someone know why they exist?

I implemented it because @sbastrakov needed it for the complex implementation some time ago: 
https://github.com/alpaka-group/alpaka/pull/1248

**Comment by bernhardmgruber** [comment]
> If you have your own types, outside of alpaka, why would you want to use alpaka::math:: on them? Would you not have your own sqrt in your namespace and call that via sqrt(varOfMyType) (or by qualified name)?

Templates. You might want to write an alpaka device side function that you can call on `float`, `double`, `MyBigNumber`, `MyComplex`, `__half ` or `bfloat16`.

**Comment by sbastrakov** [comment]
The idea is, one always calls `alpaka::math::func(acc, ...)` for any math function inside kernel. If there is a custom implementation, it should be fallen back to via ADL

**Comment by jkelling** [comment]
If this is the only idea, and there is nothing about compiler intrinsics, then maybe we should disable these fallback for `float` `double` `long double` to cause a compiler error when they are called accidentally.

**Comment by bernhardmgruber** [comment]
> ... maybe we should disable these fallback for float double long double to cause a compiler error when they are called accidentally.

Fine for me, send a PR!

**Comment by jkelling** [comment]
The shortest check would be using `std::is_floating_point` which specifically checking for `float`, `double` and `long double`. 

However, a user might then specialize this trait for their type for some reason and thereby inadvertently kick them out of this ADL fallback.

On the other hand, to quote [cpprefeence.com](https://en.cppreference.com/w/cpp/types/is_floating_point)
```
The behavior of a program that adds specializations for is_floating_point [...] is undefined. 
```
Are we comfortable relying on users not invoking undefined behavior?

**Comment by sbastrakov** [comment]
I am now again getting issues with this ADL stuff for my complex numbers PR. I think it is implemented maybe not in the best way. Could we just do the following instead, e.g. for `Abs`?

```cpp
// abs/Traits.hpp

//! The abs trait.
template<typename T, typename TArg, typename TSfinae = void>
struct Abs
{
    ALPAKA_FN_HOST_ACC auto operator()(T const& ctx, TArg const& arg)
    {
        alpaka::ignore_unused(ctx);
        
        // Only this line is added to automatically use std:: versions in this general specialization
        using std::abs;

        // This is an ADL call. If you get a compile error here then your type is not supported by the
        // backend and we could not find abs(TArg) in the namespace of your type.
        return abs(arg);
    }
};

// abs/AbsStdLib.hpp

(Still define AbsStdLib but not specialize Abs for it, so that the general version is used) 

// abs/AbsUniformCudaHipBuiltIn.hpp - unchanged
```

**Comment by sbastrakov** [comment]
With such a change the unit tests in #1336 seem to compile for both real and complex types. So i would think it doesn't break anything. I can make a separate PR for it to see if it works for all combinations in CI

**Comment by j-stephan** [comment]
Is this still an issue?

**Comment by sbastrakov** [comment]
The part that I was talking about is now solved afaik. I am not sure it's exactly same as the original post in this issue.

**Comment by j-stephan** [comment]
@jkelling @bernhardmgruber Do you know if this is solved?

**Comment by bernhardmgruber** [comment]
No idea, sorry.

**Comment by jkelling** [comment]
I think #1457 solved this.

**Comment by sbastrakov** [comment]
I think there is a good chance it did. So imo it's reasonable to consider this issue solved.

**Comment by j-stephan** [comment]
Thanks! Feel free to reopen if this should appear again.
